services/scribe/backfill/fetcher.go

```diff
@@ -0,0 +1,210 @@
+func (f *RangeFilter) Start(ctx context.Context) error {
+	var wg sync.WaitGroup
+
+	defer f.closeOnDone()
+
+	for {
+		select {
+		case <-ctx.Done():
+			if ctx.Err() != nil {
+				LogEvent(WarnLevel, "could not finish filtering range", LogData{"ca": f.contractAddress, "sh": f.startBlock.String(), "eh": f.endBlock.String(), "cid": &f.chainConfig.ChainID})
+				return fmt.Errorf("could not finish filtering range: %w", ctx.Err())
+			}
+
+			return nil
+		default:
+			chunks := f.GetChunkArr()
+
+			if len(chunks) == 0 {
+				wg.Wait()
+				return nil
+			}
+			logs, err := f.FilterLogs(ctx, chunks)
+			if err != nil {
+				return fmt.Errorf("could not filter logs: %w", err)
+			}
+
+			wg.Add(1)
+			go func(logs []types.Log) {
+				defer wg.Done()
+				f.logs <- logs
+			}(logs)
+			LogEvent(InfoLevel, "Contract backfill chunk completed", LogData{"ca": f.contractAddress, "sh": chunks[0].MinBlock(), "eh": chunks[0].MaxBlock(), "cid": &f.chainConfig.ChainID})
+		}
+	}
+}
+
+// FilterLogs safely calls FilterLogs with the filtering implementing a backoff in the case of
+// rate limiting and respects context cancellation.
+//
+// nolint:cyclop
+func (f *RangeFilter) FilterLogs(ctx context.Context, chunks []*util.Chunk) ([]types.Log, error) {
+	b := &backoff.Backoff{
+		Factor: 2,
+		Jitter: true,
+		Min:    minBackoff,
+		Max:    maxBackoff,
+	}
+
+	attempt := 0
+	timeout := time.Duration(0)
+
+	// for logging purposes
+	startHeight := chunks[0].StartBlock.Uint64()
+	endHeight := chunks[len(chunks)-1].EndBlock.Uint64()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, fmt.Errorf("could not finish filtering logs: %w", ctx.Err())
+		case <-time.After(timeout):
+			attempt++
+
+			if attempt > retryTolerance {
+				return nil, fmt.Errorf("maximum number of filter attempts exceeded")
+			}
+
+			res, err := GetLogsInRange(ctx, f.backend, f.contractAddress, uint64(f.chainConfig.ChainID), chunks)
+			if err != nil {
+				timeout = b.Duration()
+				LogEvent(WarnLevel, "Could not filter logs for range, retrying", LogData{"sh": startHeight, "ca": f.contractAddress, "eh": endHeight, "cid": &f.chainConfig.ChainID, "e": err})
+
+				continue
+			}
+
+			var logs []types.Log
+			itr := res.Iterator()
+			for !itr.Done() {
+				select {
+				case <-ctx.Done():
+					return nil, fmt.Errorf("could not finish filtering logs: %w", ctx.Err())
+				default:
+					_, resLogChunk := itr.Next()
+
+					if resLogChunk == nil || len(*resLogChunk) == 0 {
+						LogEvent(WarnLevel, "empty subchunk", LogData{"sh": startHeight, "ca": f.contractAddress, "cid": &f.chainConfig.ChainID, "eh": endHeight})
+						continue
+					}
+					logsChunk := *resLogChunk
+
+					logs = append(logs, logsChunk...)
+				}
+			}
+
+			return logs, nil
+		}
+	}
+}
+
+// Drain fetches empties the log chan. For use once the doneChan is emitted.
+func (f *RangeFilter) Drain(ctx context.Context) (filteredLogs []types.Log, err error) {
```

Comment:
does this need to be called by external users? If so why? If not, let's make it private